Update to @hapi/joi v15

codeandcats opened the pull request:
### Summary
- Depends on latest version of joi ([@hapi/joi](https://www.npmjs.com/package/@hapi/joi) v15)
- Had to update a couple tests which were failing due to changes in joi
- Closes #21
- Closes #8

### Notes
- Still requires joi as a peer. I can make it a normal dependency as part of this PR if you like?
- ~~Although the codebase is building fine and tests are passing, I noticed one error in my _IDE_ that I had trouble groking... @laurence-myers would you be able to take a quick look at this and update this branch? Also I'm not sure why this error doesn't break the build... 🤔~~ (addressed)
<img width="1161" alt="Screen Shot 2019-07-17 at 10 53 10 pm" src="https://user-images.githubusercontent.com/6035934/61378160-19308500-a8e8-11e9-8f16-12a148a63a73.png">

codeandcats commented on src/Validator.ts:
```diff
@@ -8,7 +8,7 @@ export class Validator {
     }
 
-    validate<T>(target: T, options?: ValidationOptions): ValidationResult<T> {
+    validate<T extends { constructor: Function }>(target: T, options?: ValidationOptions): ValidationResult<T> {
```
Had to add this, otherwise line further down didn't know about the `constructor` property.

codeandcats commented on src/constraints/any.ts:
```diff
@@ -24,7 +24,7 @@ export function AnySchema(schemaBuilder?: (schema: Schema) => Schema): TypedProp
 export function Concat(schema: Schema): TypedPropertyDecorator<AllowedPropertyTypes> {
     return constraintDecorator<AllowedPropertyTypes>((existingSchema: Schema) => {
-        return existingSchema.concat(schema);
+        return existingSchema.concat(schema as any);
```
🤷‍♂ TBH, I don't really understand this one. Imo, looking at the code, `concat` should accept a `Schema` instance, which is exactly what `schema` is. 🤔 IDK, I think this is legit but if you can find a better alternative to `any` here I'm all ears. 😃

codeandcats commented on src/constraints/any.ts:
```diff
@@ -58,7 +58,7 @@ export function Empty(schema: any): TypedPropertyDecorator<AllowedPropertyTypes>
 export function Error(err: Error | ValidationErrorFunction): TypedPropertyDecorator<AllowedPropertyTypes> {
     return constraintDecorator<AllowedPropertyTypes>((schema: Schema) => {
-        return schema.error!(err);
+        return schema.error(err);
```
Removed unnecessary bang

codeandcats commented on src/constraints/any.ts:
```diff
@@ -213,11 +213,11 @@ export const Equal = Valid;
+export function When(
     ref: string | Reference,
-    options: WhenOptions<TWhen>,
+    options: WhenOptions,
```
`WhenOptions` is no longer generic (#8)

codeandcats commented on test/unit/constraints/string.test.ts:
```diff
@@ -208,8 +194,8 @@ describe('String constraints', () => {
-            ['monkey@see.com', 'monkey@do', 'howdy+there@pardner.co.kr'],
-            ['123.com'],
+            ['monkey@see.com', 'howdy+there@pardner.co.kr'],
+            ['123.com', 'monkey@do'],
```
Seems like the email validator expects a TLD now.

codeandcats commented on test/unit/constraints/string.test.ts:
```diff
@@ -357,8 +343,8 @@ describe('String constraints', () => {
                 },
-                ['127.0.0.1/24' /*"2001:db8:abcd:8000::/50"*/], // bug in Joi? doesn't like IPv6 CIDR
-                ['127.0.0.1', '2001:0db8:0000:0000:0000:ff00:0042:8329', '2001:db8:abcd:8000::/50'],
+                ['127.0.0.1/24', '2001:db8:abcd:8000::/50'],
```
They seem to have fixed this for IPv6 😄

codeandcats commented on test/unit/constraints/string.test.ts:
```diff
@@ -464,7 +450,7 @@ describe('String constraints', () => {
             () => {
                 class MyClass {
-                    @Regex(/test/g)
+                    @Regex(/test/)
```
Joi's regex validator doesn't like the regex having the global flag. Joi seems to treat it like it is global regardless.

codeandcats commented on test/unit/constraints/string.test.ts:
```diff
@@ -473,8 +459,8 @@ describe('String constraints', () => {
             },
-            ['test123', '123test', '123test123'],
-            ['abcd', 'TEST'],
+            ['test', 'test123', '123test', '123test123'],
```
Nothing special here, just added `'test'` for more clarity.

laurence-myers commented:
Awesome work! I'll have a detailed look later.

CircleCI failed it's complaining about the presence of package-lock.json, and that yarn.lock hasn't been updated.

Re: the TS message, it's saying that `unknown` and `object` are not compatible. Does `typeConstraintDecorator()` need a type parameter? Otherwise, this might be a bug in the type-safe decorator type, it might need to ignore object properties with `unknown` types. What version of TS is your VS Code using? If it's different to tsdv-joi's version, that might explain why the build is not failing, but the error comes up in your editor.

codeandcats commented:
Thanks for the quick feedback.

One day I'll remember to use yarn hehe. I've removed the `package-lock.json` and updated `yarn.lock` now.

I've explicitly set the type parameter for that `typeConstraintDecorator` call like we do everywhere else.

You were right about the IDE using a different TypeScript version, I thought I had it set to always use the version in the workspace but I hadn't. 👌

Also, I took the liberty of merging in #43 as Joi v15 needed a more modern node which was causing CI to fail for this branch.

codeandcats commented:
Closing this PR since #54 is a major rewrite and will be easier to upgrade joi in a new PR than to try and rebase/merge into this branch.
